algorithms/linkedlist/copy_random_pointer.py

```diff
@@ -4,35 +4,90 @@
 
 Return a deep copy of the list.
 """
-# from collections import defaultdict
-
-# TODO: This requires to be rewritten -- commenting it out for now
-# class Solution0:
-  # @param head, a RandomListNode
-  # @return a RandomListNode
-  # def copyRandomList(self, head):
-  #   dic = dict()
-  #   m = n = head
-  #   while m:
-  #       dic[m] = RandomListNode(m.label)
-  #       m = m.next
-  #   while n:
-  #       dic[n].next = dic.get(n.next)
-  #       dic[n].random = dic.get(n.random)
-  #       n = n.next
-  #   return dic.get(head)
+from collections import defaultdict
+from unittest import TestCase
+
+
+class RandomListNode(object):
+    def __init__(self, label):
+        self.label = label
+        self.next = None
+        self.random = None
+
+
+class Solution0:
+    def copy_random_list(self, head):
+        """
+        :type head: RandomListNode
+        :rtype: RandomListNode
+        """
+        dic = dict()
+        m = n = head
+        while m:
+            dic[m] = RandomListNode(m.label)
+            m = m.next
+        while n:
+            dic[n].next = dic.get(n.next)
+            dic[n].random = dic.get(n.random)
+            n = n.next
+        return dic.get(head)
+
+
 #
 #
-# class Solution1:  # O(n)
-#   @param head, a RandomListNode
-#   @return a RandomListNode
-  # def copyRandomList(self, head):
-  #   copy = defaultdict(lambda: RandomListNode(0))
-  #   copy[None] = None
-  #   node = head
-  #   while node:
-  #       copy[node].label = node.label
-  #       copy[node].next = copy[node.next]
-  #       copy[node].random = copy[node.random]
-  #       node = node.next
-  #   return copy[head]
+class Solution1:  # O(n)
+    def copy_random_list(self, head):
+        """
+        :type head: RandomListNode
+        :rtype: RandomListNode
+        """
+        copy = defaultdict(lambda: RandomListNode(0))
+        copy[None] = None
+        node = head
+        while node:
+            copy[node].label = node.label
+            copy[node].next = copy[node.next]
+            copy[node].random = copy[node.random]
+            node = node.next
+        return copy[head]
+
+
+class TestSuite(TestCase):
```

Comment:
could you move this to [tests folder](https://github.com/keon/algorithms/blob/master/tests/test_linkedlist.py)?